### ISSUE TITLE
How to use with net5.0-windows builds?

### ISSUE THREAD

**Issue by Noemata** [mveril/WinRTInitializeWithWindow#1]
I tried incorporating this into a net5.0-windows Wpf app, but had no luck.  Any ideas?

**Comment by Noemata** [comment]
Never mind.  Turns out there's an easy way to do this using your code and CsWinRT:

```
using System;
using System.Windows;
using System.Windows.Interop;
using System.Runtime.InteropServices;

namespace WinRT.InitializeWithWindow
{
    /// <summary>
    /// <see langword="static"/> <see langword="class"/> that contain the WPF <see cref="Window"/> extension method
    /// </summary>
    public static class WPFWindowExtension
    {
        /// <summary>
        /// COM Import of the native <c>IInitializeWithWindow</c> <see langword="interface"></see> <see href="https://devblogs.microsoft.com/oldnewthing/20190412-00/?p=102413">Related blog post</see> or <seealso href="https://docs.microsoft.com/windows/uwp/design/shell/tiles-and-notifications/secondary-tiles-desktop-pinning">this C# example</seealso>
        /// </summary>
        [ComImport]
        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        [Guid("3E68D4BD-7135-4D10-8018-9FB6D9F33FA1")]
        public interface IInitializeWithWindow
        {
            void Initialize(IntPtr hwnd);
        }

        /// <summary>
        /// Extension method to initialize WinRT object with this <see cref="Window"/>
        /// </summary>
        /// <param name="window"></param>
        /// <param name="winRTObj"></param>
        public static void InitializeWinRTChild(this Window window, object winRTObj)
        {
            var withWindow = winRTObj.As<IInitializeWithWindow>(); ;
            withWindow.Initialize(new WindowInteropHelper(window).Handle);
        }

        /// <summary>
        /// Method to initialize WinRT object.
        /// </summary>
        /// <param name="winRTObj"></param>
        /// <param name="window"></param>
        public static void InitializeWinRTChild(object winRTObj, Window window = null)
        {
            var withWindow = winRTObj.As<IInitializeWithWindow>(); ;
            withWindow.Initialize(new WindowInteropHelper(window == null ? Application.Current.MainWindow : window).Handle);
        }
    }
}
```

**Comment by mveril** [comment]
I never took the time to update it after the [CsWinRT](https://github.com/microsoft/CsWinRT) update  that came with .NET 5 because now it's very easy to use the interface directly unlike with the .NET 5 preview. You can see the difference on the .NET 5 and WinUI branch of this repo. So this repo is a little useless now.
What do you think about it? It is interesting to update it for .NET 5 (and also WinUI 3) or I should mark this repo as depreciated?

**Comment by Noemata** [comment]
@mveril, your packaging of IInitializeWithWindow and extension method does make it somewhat simpler to use.  I liked it!  I also added one new method as you may have noticed.  What's still unresolved is getting a definition for IAsyncAction so awaitable calls from UWP can be used.  I've yet to get an answer from Microsoft on this.

**Comment by mveril** [comment]
@Noemata thank you for your interest about this project ! Folowing your comment I working on an update of this project which support .NET 5 and also WinUI 3.
I don't realy understand your issue about [IAsyncAction](https://docs.microsoft.com/uwp/api/Windows.Foundation.IAsyncAction). If you use a NET 5 Windows 10 TFM (see the [Calling Windows APIs in .NET5 blog post](https://blogs.windows.com/windowsdeveloper/2020/09/03/calling-windows-apis-in-net5)) all the interop are included so normally there is no problem with [IAsyncAction](https://docs.microsoft.com/uwp/api/Windows.Foundation.IAsyncAction).

**Comment by Noemata** [comment]
@mveril, I don't understand it either.  But the problem is there.  Have a look at: https://github.com/Noemata/AssetInstaller

I had to add an implementation for IAsyncAction to get things working.  If you know of a workaround, please let me know.  I think this is something that was working and may have been broken by recent releases of .Net 5.

**Comment by mveril** [comment]
@Noemata I made a pull request on your repo that solve your IAsyncAction problem you don't need to import the Windows 10 SDK manually on .NET 5. The SDK Contracts package is not compatible with this version of .NET

**Comment by Noemata** [comment]
Silly me.  Thank you very much.   It's been ages since I did anything with WPF.  Are XAML Islands supported in .Net 5 WPF builds?  I found the docs around all of this hard to follow given all the recent changes.  Thanks again.

**Comment by Noemata** [comment]
Answering my own question, XAML Islands are not supported in .NET 5 and will not be available until WinUI 3.x is out (post WinUI 3.0).

**Comment by mveril** [comment]
Hello @Noemata
I recently discovered [here](https://github.com/microsoft/WindowsAppSDK/issues/1063#issuecomment-881023368)  that [C#/WinRT 1.3](https://github.com/microsoft/CsWinRT/releases/tag/1.3.0.210617.2) now provides projections for [COM Interop APIs](https://docs.microsoft.com/windows/apps/desktop/modernize/winrt-com-interop-csharp) including [InitializeWithWindow](https://github.com/microsoft/CsWinRT/blob/4560924608f47e360d4dfc1512ded7c4f530d873/src/cswinrt/strings/ComInteropHelpers.cs#L71). So, I highly recommend you to use it.
Here a practical example.
``` csharp
var dlg = new MessageDialog("Hellow World",$"UWP {nameof(MessageDialog)}") ;
var hwind = new WindowInteropHelper(this).Handle;
WinRT.Interop.InitializeWithWindow.Initialize(dlg, hwind);
await dlg.ShowAsync();
```
Due to the presence of official interop APIs as I wrote 6952bab, I decided to never provide support for .NET 5 and WinUI 3. Since the interop APIs are easy enough to use and does not require to import COM interfaces by hand.
Best regards
